### PR TITLE
fix(HACBS-1821): update fbc-validation to correctly validate fbc fragment

### PR DESCRIPTION
**Pull request by jsztuka**
Fixes vital checks for FBC fragment

- presence of opm binary
- presence of grpc_health_probe
- runs opm validate on catalog
- makes sure there is only one opm.package present

### CODE REVIEW

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 1]
```diff
@@ -57,32 +63,35 @@ spec:
...
-        if ! ./opm render confdir | jq -en 'reduce (inputs | select(.schema == "olm.package")) as $obj (0; .+1) == 1'; then
+        if ! opm render ."${conffolder}" | jq -en 'reduce (inputs | select(.schema == "olm.package")) as $obj (0; .+1) == 1'; then
           echo "!FAILURE! - more than one olm.packages are not permitted in a FBC fragment"
           failure_num=`expr $failure_num + 1`
           TESTPASSED=false
         fi
         if [ $TESTPASSED == false ]; then
           HACBS_ERROR_OUTPUT="$(make_result_json -r FAILURE -f $failure_num -s `expr $check_num - $failure_num`)"
           echo "${HACBS_ERROR_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
+          cd ..
```
Comment (hongweiliu17) [inline task/fbc-validation/0.1/fbc-validation.yaml:90]: We may delete the file out of the  `if..else..` so we can have only one `cd & rm -fr`, but it is not a big deal.

**Review by hongweiliu17 — APPROVED** [review]

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 1 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml:90]: Thats a good point. Thanks!

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 2]
```diff
@@ -34,19 +37,22 @@ spec:
         #!/usr/bin/env bash
         source /utils.sh
         ### Try to extract binaries with configs > check binaries functionality > check opm validate ###
-        if [ ! -f ../sanity-inspect-image/image_inspect.json ] || [ -z $(cat ../sanity-inspect-image/image_inspect.json) ]; then
+
+        if [[ ! -f ../sanity-inspect-image/image_inspect.json ]] || [[ -z $(cat ../sanity-inspect-image/image_inspect.json) ]]; then
```
Comment (hongweiliu17) [inline task/fbc-validation/0.1/fbc-validation.yaml:41]: See https://github.com/redhat-appstudio/build-definitions/pull/326#discussion_r1115431258, they are similar

**Review by hongweiliu17 — APPROVED** [review]

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 3]
```diff
@@ -34,19 +37,22 @@ spec:
...
           echo "File ../sanity-inspect-image/image_inspect.json is not generated correctly, please check HACBS_TEST_OUTPUT of task sanity-inspect-image"
           HACBS_TEST_OUTPUT="$(make_result_json -r ERROR -t 'File ../sanity-inspect-image/image_inspect.json is not generated correctly, please check HACBS_TEST_OUTPUT of task sanity-inspect-image!')"
           echo "${HACBS_TEST_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
           exit 0
         fi
 
         conffolder=$(cat ../sanity-inspect-image/image_inspect.json | jq -r '.Labels ."operators.operatorframework.io.index.configs.v1"')
+        image_with_digest=$(params.IMAGE_URL)@$(params.IMAGE_DIGEST)
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:49]: Why this change?

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 4]
```diff
@@ -34,19 +37,22 @@ spec:
...
 
-        image_with_digest="$(params.IMAGE_URL)@$(params.IMAGE_DIGEST)"
+        #make temporary folder for image content which will be deleted at the end of this task
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:51]:
```suggestion
        # make temporary folder for image content which will be deleted at the end of this task
```

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 5]
```diff
@@ -34,19 +37,22 @@ spec:
...
+        mkdir to-delete
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:52]: Could we name it like catalog-fragment-content to keep semantic of what's inside?

Comment (MartinBasti) [reply task/fbc-validation/0.1/fbc-validation.yaml:52]: I see working dir is a workspace, shouldn't, why do we need copy content to workspace volume? Couldn't we just use local dir or emtpyDir? Copying to shared mount will make it much slower, we don't need it

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 6]
```diff
@@ -86,3 +91,4 @@ spec:
           HACBS_TEST_OUTPUT="$(make_result_json -r SUCCESS -s $check_num)"
           echo "${HACBS_TEST_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
         fi
+        cd .. && rm -fr to-delete
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:94]: bonus pints for making `to-delete` a constant instead of copy-pasting the same name 3 times

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 7]
```diff
@@ -57,24 +63,23 @@ spec:
...
         fi
-        if [ ! -f "grpc_health_probe" ]; then
-          echo "!FAILURE! - grpc_health_probe binary check failed"
+        if [ ! $(find . -name grpc_health_probe ) ]; then
+          echo "!FAILURE! - grpc_health_probe binary presence check failed"
           failure_num=`expr $failure_num + 1`
           TESTPASSED=false
         fi
-        if ! ./opm validate confdir; then
+        if ! opm validate ."${conffolder}"; then
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:77]: Did we validate that conffolder is non-empty to avoid weird errors?

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 8]
```diff
@@ -34,19 +37,22 @@ spec:
...
-        if [ ! -f ../sanity-inspect-image/image_inspect.json ] || [ -z $(cat ../sanity-inspect-image/image_inspect.json) ]; then
+
+        if [[ ! -f ../sanity-inspect-image/image_inspect.json ]] || [[ -z $(cat ../sanity-inspect-image/image_inspect.json) ]]; then
           echo "File ../sanity-inspect-image/image_inspect.json is not generated correctly, please check HACBS_TEST_OUTPUT of task sanity-inspect-image"
           HACBS_TEST_OUTPUT="$(make_result_json -r ERROR -t 'File ../sanity-inspect-image/image_inspect.json is not generated correctly, please check HACBS_TEST_OUTPUT of task sanity-inspect-image!')"
           echo "${HACBS_TEST_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
           exit 0
         fi
 
         conffolder=$(cat ../sanity-inspect-image/image_inspect.json | jq -r '.Labels ."operators.operatorframework.io.index.configs.v1"')
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:48]: IMO we should check if this command failed and print appropriate error message

**Review by MartinBasti — CHANGES_REQUESTED** [review]

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 3 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml:49]: Lets put back the quotes, don't know why they were stripped.

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 5 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml:52]: We are using workspace same as from previous step to get: 
../sanity-inspect-image/image_inspect.json

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 7 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml:85]: Check added.

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 5 above)
Comment (MartinBasti) [reply task/fbc-validation/0.1/fbc-validation.yaml]: yes, for `../sanity-inspect-image/image_inspect.json` it's ok, but why put there some temp data from oc extract?

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 2 above)
Comment (MartinBasti) [reply task/fbc-validation/0.1/fbc-validation.yaml:42]: uff, sorry I haven't noticed :D so change it then

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 9]
```diff
@@ -32,21 +35,39 @@ spec:
...
+        if [ $? -ne 0 ]; then
+          echo "Could not get labels from sanity-inspect-image/image_inspect.json, make sure the file exists and contains this label: operators.operatorframework.io.index.configs.v1."
+          HACBS_TEST_OUTPUT="$(make_result_json -r FAILURE -f 1)"
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:52]:
```suggestion
          HACBS_TEST_OUTPUT="$(make_result_json -r ERROR)"
```

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 5 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml]: Changed that to /tmp

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 9 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml]: Updated

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 2 above)
Comment (MartinBasti) [reply task/fbc-validation/0.1/fbc-validation.yaml:42]: @jsztuka so can we apply the same fix here?

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 10]
```diff
@@ -32,49 +35,67 @@ spec:
...
           echo "${HACBS_TEST_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
           exit 0
         fi
 
+        if [ -z "$(ls -A .$conffolder)" ]; then
+          echo "$conffolder is empty, missing catalog file."
+          HACBS_TEST_OUTPUT="$(make_result_json -r ERROR)"
+          echo "${HACBS_TEST_OUTPUT}" | tee "$(results.HACBS_TEST_OUTPUT.path)"
+          popd
+          rm -fr /tmp/*
```
Comment (MartinBasti) [inline task/fbc-validation/0.1/fbc-validation.yaml:74]: the same here

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 2 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml]: We do, updated.

File: task/fbc-validation/0.1/fbc-validation.yaml [hunk 11]
```diff
@@ -86,3 +107,5 @@ spec:
           HACBS_TEST_OUTPUT="$(make_result_json -r SUCCESS -s $check_num)"
           echo "${HACBS_TEST_OUTPUT}" | tee $(results.HACBS_TEST_OUTPUT.path)
         fi
+        popd
+        rm -fr /tmp/*
```
[the remark this answers is not in the document]
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml]: Alright, /tmp content will not be removed in this task.

File: task/fbc-validation/0.1/fbc-validation.yaml (hunk 10 above)
Comment (jsztuka) [reply task/fbc-validation/0.1/fbc-validation.yaml]: Same as above.

**Review by MartinBasti — APPROVED** [review]
